cmd/juju/action/call.go

```diff
@@ -0,0 +1,310 @@
+Params may be in a yaml file which is passed with the --params option, or they
+may be specified by a key.key.key...=value format (see examples below.)
+
+Params given in the CLI invocation will be parsed as YAML unless the
+--string-args option is set.  This can be helpful for values such as 'y', which
+is a boolean true in YAML.
+
+If --params is passed, along with key.key...=value explicit arguments, the
+explicit arguments will override the parameter file.
+
+Examples:
+
+    juju call mysql/3 backup --wait
+    juju call mysql/3 backup
+    juju call mysql/leader backup
+    juju show-action-output <ID>
+    juju call mysql/3 backup --params parameters.yml
+    juju call mysql/3 backup out=out.tar.bz2 file.kind=xz file.quality=high
+    juju call mysql/3 backup --params p.yml file.kind=xz file.quality=high
+    juju call sleeper/0 pause time=1000
+    juju call sleeper/0 pause --string-args time=1000
+`
+
+// SetFlags offers an option for YAML output.
+func (c *callCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ActionCommandBase.SetFlags(f)
+	c.out.AddFlags(f, "yaml", output.DefaultFormatters)
+	f.Var(&c.paramsYAML, "params", "Path to yaml-formatted params file")
+	f.BoolVar(&c.parseStrings, "string-args", false, "Use raw string values of CLI args")
+	f.Var(&c.wait, "wait", "Wait for results, with optional timeout")
+}
+
+func (c *callCommand) Info() *cmd.Info {
+	return jujucmd.Info(&cmd.Info{
+		Name:    "call",
+		Args:    "<unit> [<unit> ...] <action name> [key.key.key...=value]",
+		Purpose: "Queue an action for execution.",
+		Doc:     callDoc,
+	})
+}
+
+// Init gets the unit tag(s), action name and action arguments.
+func (c *callCommand) Init(args []string) (err error) {
+	for _, arg := range args {
+		if names.IsValidUnit(arg) || validLeader.MatchString(arg) {
+			c.unitReceivers = append(c.unitReceivers, arg)
+		} else if nameRule.MatchString(arg) {
+			c.actionName = arg
+			break
+		} else {
+			return errors.Errorf("invalid unit or action name %q", arg)
+		}
+	}
+	if len(c.unitReceivers) == 0 {
+		return errors.New("no unit specified")
+	}
+	if c.actionName == "" {
+		return errors.New("no action specified")
+	}
+
+	// Parse CLI key-value args if they exist.
+	c.args = make([][]string, 0)
+	for _, arg := range args[len(c.unitReceivers)+1:] {
+		thisArg := strings.SplitN(arg, "=", 2)
+		if len(thisArg) != 2 {
+			return errors.Errorf("argument %q must be of the form key...=value", arg)
+		}
+		keySlice := strings.Split(thisArg[0], ".")
+		// check each key for validity
+		for _, key := range keySlice {
+			if valid := nameRule.MatchString(key); !valid {
+				return errors.Errorf("key %q must start and end with lowercase alphanumeric, "+
+					"and contain only lowercase alphanumeric and hyphens", key)
+			}
+		}
+		// c.args={..., [key, key, key, key, value]}
+		c.args = append(c.args, append(keySlice, thisArg[1]))
+	}
+	return nil
+}
+
+func (c *callCommand) Run(ctx *cmd.Context) error {
+	if err := c.ensureAPI(); err != nil {
```

Comment:
This is kind of a problem since we close api later on but this method can still return it. I do prefer commands having function to obtain new API rather than caching the api.

Reply:
Yeah, this is all existing, just a straight copy. Will try to clean up as work progresses